chains/substrate/connection.go

```diff
@@ -21,10 +22,17 @@ type Connection struct {
 	meta        *types.Metadata
 	genesisHash types.Hash
 	key         *signature.KeyringPair
+	metaLock    sync.RWMutex
 }
 
 func NewConnection(url string, key *signature.KeyringPair) *Connection {
-	return &Connection{url: url, key: key}
+	return &Connection{url: url, key: key, metaLock: sync.RWMutex{}}
+}
+
+func (c *Connection) getMetadata() *types.Metadata {
+	c.metaLock.Lock()
+	defer c.metaLock.Unlock()
+	return c.meta
```

Comment:
You're returning a pointer which nullifies the point of having a lock as you still have concurrent access to shared data. You need to return a value (ie. a copy of the data). `Connection.meta` could be a value instead a pointer since we only really want copies of it.

This could also be optimized by using `RLock()` and `RUnlock()` as these are only reads.